File: 141.linked-list-cycle.java
import java.util.HashMap;

/*
 * @lc app=leetcode id=141 lang=java
 *
 * [141] Linked List Cycle
 */

// @lc code=start
/**
 * Definition for singly-linked list. class ListNode { int val; ListNode next;
 * ListNode(int x) { val = x; next = null; } }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        HashMap<ListNode, Integer> map = new HashMap<>();
        while (head != null) {
            map.put(head, map.getOrDefault(head, 0) + 1);
            if (map.get(head) > 1) {
                return true;
            }
            head = head.next;
        }
        return false;
    }
}
// @lc code=end

File: 141.linked-list-cycle.py
#
# @lc app=leetcode id=141 lang=python3
#
# [141] Linked List Cycle
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head:
            return False

        slow, fast= head, head.next
        while slow != fast:
            if fast and fast.next:
                slow=slow.next
                fast=fast.next.next
            else:
                return False
        
        return True  
# @lc code=end

